packages/@lwc/engine-core/src/framework/rendering.ts

```diff
@@ -618,16 +617,55 @@ export function allocateChildren(vnode: VCustomElement, vm: VM) {
             );
         }
     }
+
+    // flatten vfragment nodes and mark dynamic
+    const { flattened, flattenedChildren } = flattenFragmentChildren(children);
```

Comment:
Note that in the use case where we're rerendering from a `VFragment` to a `null`, the children will not be flattened or marked as dynamic children. In that situation, if the number of elements changes, the elements are not rerendered correctly because we go down the static diffing algo instead of the dynamic one.

```javascript
// array marked as dynamic, vfragment children are flattened
[div, vfragment, div]

// array marked as static, nothing tells us we need to mark this dynamic
[div, null, div]
```

Using the karma test component `parentWithNamedSlot`, this is what I saw for this use case:
```javascript
// with condition === true
static slot header
conditional slot content
static slot footer
Default slot content

// with condition === false
static slot header
static slot footer
static slot footer // note that this node is not properly unrendered
Default slot content
```

One way to solve for this is to replace the `null` with an empty `VFragment`. This lets us know that even though the nodes are empty, we still need to go down the dynamic path to properly unrender the correct elements:
```javascript
// array marked as dynamic, vfragment children are flattened
[div, vfragment([...]), div]

// array marked as dynamic, still renders as [div, null, div]
[div, vfragment([]), div]
```